---
title: "Analysis of Adaptive Immune Receptor Repertoire Sequencing data with LymphoSeq2"
author: 
 - "Elena Wu"
 - "Shashidhar Ravishankar"
 - "David Coffey"
date: "`r Sys.Date()`"
output: 
    html_document:
        toc: true
        toc_float: true
        theme: cosmo
        highlight: default
        df_print: paged
vignette: >
  %\VignetteIndexEntry{LymphoSeq2}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

Adaptive Immune Receptor Repertoire Sequencing (AIRR-seq) provides a unique opportunity to interrogate the adaptive
immune repertoire under various clinical conditions. The utility offered by this technology has quickly garnered interest
from a community of clinicians and researchers investigating the immunological landscapes of a large spectrum of health
and disease states. LymphoSeq2 is a toolkit that allows users to import, manipulate and visualize AIRR-Seq data from various 
AIRR-Seq assays such as Adaptive ImmunoSEQ and BGI-IRSeq, with support for 10X VDJ sequencing coming soon. The platform 
also supports the importing of AIRR-seq data processed using the MiXCR pipeline. The vignette highlights some of the key features of LymphoSeq2.

# Importing data

The function `readImmunoSeq` imports AIRR-seq receptor files from Adaptive ImmunoSEQ assay as well well as BGI-IRSeq assay. The sequences can be (.tsv) files processed using one of the three following platforms: Adaptive Biotechnologies ImmunoSEQ analyzer, BGI IR-SEQ iMonitor platform, and the MiXCR pipeline for AIRR-seq data analysis. The function has the ability to identify file type based on the headers provided in the (.tsv) file, accordingly the data is transformed into a format that is compatible AIRR-Community guidelines  (https://github.com/airr-community/airr-standards). 

To explore the features of LymphoSeq, this package includes 2 example data sets. The first is a data set of T cell receptor beta (TCRB) sequencing from 10 blood samples acquired serially from a single patient who underwent a bone marrow transplant (Kanakry, C.G., et al. JCI Insight 2016;1(5):pii: e86252). The second, is a data set of B cell receptor immunoglobulin heavy (IGH) chain sequencing from Burkitt lymphoma tumor biopsies acquired from 10 different individuals (Lombardo, K.A., et al. Blood Advances 2017 1:535-544). To improve performance, both data sets contain only the top 1,000 most frequent sequences. The complete data sets are publicly available through Adapativesâ€™ immuneACCESS portal. As shown in the example below, you can specify the path to the example data sets using the command 

```{r data}
system.file("extdata", "TCRB_sequencing", package = "LymphoSeq2") #For the TCRB files
system.file("extdata", "IGH_sequencing", package = "LymphoSeq2") #For the IGH files.
```

`readImmunoSeq` can take as input, a single file name, a list of files or the path to a directory containing AIRR-seq data. The columns are renamed to follow AIRR-community guidelines based on the input file type. The function returns a tibble with individual file names set as the `repertoire_id`. The CDR3 nucleotide and amino acid sequences are denoted by the `junction` and `junction_aa` fields respectively. The counts of the CDR3 sequeneces observed, and their frequency in each individual repertoire is denoted by the `duplicate_count` and `duplicate_frequency` field respectively.

```{r readImmuno}
library(LymphoSeq2)
library(tidyverse)
study_files <- system.file("extdata", "TCRB_sequencing", package = "LymphoSeq2") 
study_table <- LymphoSeq2::readImmunoSeq(path = study_files, threads = 1)
```

Looking at the `study_table` we see a tibble with 12 columns and 9334 rows

```{r showStudyTable}
study_table
```
Since the study table is a tibble, we can use tidyverse syntax to extract a list of sample names

```{r getSamplesNames}
study_table %>% 
dplyr::pull(repertoire_id) %>% 
unique()
```


# Session info

```{r sessionInfo}
sessionInfo()
```
