#' Gene frequencies
#' 
#' Creates a data frame of VDJ gene counts and frequencies.
#' 
#' @param productive_table A tibble of productive sequences 
#' generated by the LymphoSeq function productiveSeq where the parameter 
#' aggregate is set to "nucleotide".
#' @param locus A character vector indicating which VDJ genes to include in the 
#' output.  Available options include "VDJ", "DJ", "VJ", "DJ", "V", "D", or "J".
#' @param family A Boolean value indicating whether or not family names instead 
#' of gene names are used.  If TRUE, then family names are used and if FALSE, 
#' gene names are used.
#' @return Returns a data frame with the repertoire_id names, VDJ gene name, duplicate_count, and 
#' \% frequency of the V, D, or J genes (each gene frequency should add to 
#' 100\% for each repertoire_id).
#' @examples
#' file.path <- system.file("extdata", "TCRB_sequencing", package = "LymphoSeq")
#' 
#' study_table <- readImmunoSeq(path = file.path)
#' 
#' productive_nt <- productiveSeq(study_table = study_table, aggregate = "nucleotide")
#' 
#' geneFreq(productive_nt, locus = "VDJ", family = FALSE)
#' 
#' # Create a heat map of V gene usage
#' vfamilies <- geneFreq(productive_nt, locus = "V", family = TRUE)
#' 
#' vfamilies <- reshape::cast(vfamilies, familyName ~ repertoire_ids, value = "frequencyGene", sum)
#' 
#' rownames(vfamilies) <- as.character(vfamilies$familyName)
#' 
#' vfamilies$familyName <- NULL
#' 
#' RedBlue <- grDevices::colorRampPalette(rev(RColorBrewer::brewer.pal(11, "RdBu")))(256)
#' 
#' require(pheatmap)
#' 
#' pheatmap::pheatmap(vfamilies, color = RedBlue, scale = "row")
#' 
#' # Create a word cloud of V gene usage
#' vgenes <- geneFreq(productive.nt, locus = "V", family = FALSE)
#' 
#' require(wordcloud)
#' 
#' wordcloud::wordcloud(words = vgenes[vgenes$repertoire_ids == "TRB_Unsorted_83", "geneName"], 
#'    freq = vgenes[vgenes$repertoire_ids == "TRB_Unsorted_83", "frequencyGene"], 
#' 	  colors = RedBlue)
#' 
#' # Create a cumulative frequency bar plot of V gene usage
#' vgenes <- geneFreq(productive.nt, locus = "V", family = FALSE)
#' 
#' require(ggplot2)
#' 
#' ggplot2::ggplot(vgenes, aes(x = repertoire_ids, y = frequencyGene, fill = geneName)) +
#'   geom_bar(stat = "identity") +
#'   theme_minimal() + 
#'   scale_y_continuous(expand = c(0, 0)) + 
#'   guides(fill = guide_legend(ncol = 3)) +
#'   labs(y = "Frequency (%)", x = "", fill = "") +
#'   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
#' @export
#' @import tidyverse
#' @importFrom stats aggregate
geneFreq <- function(productive_nt, locus = "V|D|J", family = FALSE) {
    if (family){
        gene_names <- productive_nt %>% 
                      select(repertoire_id, duplicate_count, v_family, j_family, d_family) %>%
                      pivot_longer(cols = contains("FamilyName"), 
                                   values_to="geneName", 
                                   names_to="geneType") %>%
                      filter(str_detect(geneName, paste("[", locus, to_lower(locus), "]", sep=""))) %>%
                      group_by(repertoire_id, geneName) %>% 
                      summarize(frequencyGene = duplicate_count/sum(duplicate_count) * 100) %>%
                     ungroup()
    } else {
        gene_names <- productive_nt %>% 
                         select(repertoire_id, duplicate_count, v_family, j_family, d_family) %>%
                         pivot_longer(cols = contains("FamilyName"), 
                                      values_to="geneName", 
                                      names_to="geneType") %>%
                         filter(str_detect(geneName, paste("[", locus, to_lower(locus), "]", sep=""))) %>%
                         group_by(repertoire_id, geneName) %>% 
                         summarize(duplicate_count = sum(duplicate_count)) %>%
                         mutate(frequencyGene = duplicate_count/sum(duplicate_count) * 100) %>%
                         ungroup()
    }
    return(gene_names)
}