#' Create phylogenetic tree
#' 
#' Create a phylogenetic tree using neighbor joining tree estimation for amino acid or 
#' junction CDR3 sequences in a list of data frames.
#' 
#' @param study_table A tibble of unproductive junction sequences or productive 
#' junction sequences generated by the LymphoSeq function productiveSeq.  
#' v_family, d_family, j_family, and duplicate_count are required columns.
#' @param repertoire_id A character vector indicating the name of the repertoire_id in the study table. 
#' @param type A character vector indicating whether "junction_aa" or "junction" sequences
#' should be compared.
#' @param layout A character vector indicating the tree layout.  Options include 
#' "rectangular", "slanted", "fan", "circular", "radial" and "unrooted".
#' @param label A Boolean indicating if the sequencing duplicate_count should be shown next to the leaves.
#' @return Returns a phylogenetic tree where each leaf represents a sequence color coded by the
#' V, D, and J gene usage.  The number next to each leaf refers to the sequence duplicate_count.  A triangle 
#' shaped leaf indicates the dominant sequence.  Refer to the ggtree Bioconductor package 
#' documentation for details on how to manipulate the tree.
#' @examples
#' file.path <- system.file("extdata", "IGH_sequencing", package = "LymphoSeq")
#' 
#' study_table <- readImmunoSeq(path = file.path)
#' 
#' productive_nt <- productiveSeq(study_table = study_table, aggregate = "junction")
#' 
#' phyloTree(study_table = productive_nt, repertoire_id = "IGH_MVQ92552A_BL", type = "junction", 
#'          layout = "rectangular")
#' 
#' phyloTree(study_table = productive_nt, repertoire_id = "IGH_MVQ92552A_BL", type = "junction_aa", 
#'          layout = "circular")
#'          
#' # Add scale and title to figure
#' library(ggtree)
#' library(ggplot2)
#' phyloTree(study_table = productive_nt, repertoire_id = "IGH_MVQ92552A_BL", type = "junction_aa", 
#'          layout = "rectangular") +
#'          ggtree::theme_tree2() +
#'          ggplot2::theme(legend.position = "right", legend.key = element_rect(colour = "white")) +
#'          ggplot2::ggtitle("Title")
#'          
#' # Hide legend and leaf labels
#' phyloTree(study_table = productive_nt, repertoire_id = "IGH_MVQ92552A_BL", type = "junction", 
#'          layout = "rectangular", label = FALSE) +
#'          ggplot2::theme(legend.position="none")
#'          
#' @export
#' @import ggtree
#' @import tidyverse
#' @import ggplot2
#' @importFrom phangorn NJ
phyloTree <- function(study_table, repertoire_id, type = "junction", layout = "rectangular", label = TRUE) {
    sample_table <- study_table %>% 
                    filter(repertoire_id == repertoire_id)
    if(nrow(study_table) < 3){
        stop("Cannot draw phlogenetic tree with less than 3 sequences.", call. = FALSE)
    }
    sample_table <- sample_table %>% 
                    mutate_all(funs(str_replace(., "", "unresolved")))
    if(type == "junction") {
        sample_table <- sample_table %>% 
                        filter(nchar(junction) >= 9)
        distance <- adist(sample_table$junction, sample_table$junction)
        colnames(distance) <- rownames(distance) <- sample_table$junction
        names <- sample_table$junction
    }
    if(type == "junction_aa") {
        sample_table <- sample_table %>% 
                        filter(nchar(junction_aa) >= 9)
        distance <- adist(sample_table$junction_aa, sample_table$junction_aa)
        colnames(distance) <- rownames(distance) <- sample_table$junction_aa
        names <- sample_table$junction_aa
    }
    tree <- phangorn::NJ(distance)
    geneFamilies <- paste(sample_table$v_family, sample_table$d_family, sample_table$j_family)
    geneFamilies <- gsub(geneFamilies, pattern = "IGH|IGL|IGK|TCRB|TCRA", replacement = "")
    geneFamilies <- gsub(geneFamilies, pattern = "unresolved", replacement = "UNR")
    tree_annotation <- tibble(names = names, 
                              duplicate_count = sample_table$duplicate_count, 
                              geneFamilies = geneFamilies, 
                              junction_aa = sample_table$junction_aa,
                              duplicate_frequency = round(sample_table$duplicate_frequency, digits = 2), 
                              dominant = c("Yes", rep("No", nrow(sample_table) - 1)))
    getPalette <- grDevices::colorRampPalette(RColorBrewer::brewer.pal(9, "Set1"))
    plot <- ggtree::ggtree(tree, layout = layout) %<+% tree_annotation+ 
            ggtree::geom_tippoint(aes_string(color = "geneFamilies", shape = "dominant"), size = 3) + 
            ggplot2::scale_color_manual(values = getPalette(length(unique(geneFamilies)))) + 
            ggplot2::guides(shape = FALSE) + 
            ggplot2::theme(legend.position = "bottom", legend.key = element_rect(colour = "white")) + 
            ggplot2::labs(color = "")
    if(label == TRUE){
        plot <- plot + ggplot2::geom_text(aes_string(label="duplicate_count"), nudge_x = 0.5, hjust = 0, size = 2.5, na.rm = TRUE, check_overlap = TRUE)
    }
    return(plot)
}